### ISSUE TITLE
ESP32 GPIO peripheral interrupts issue - manifests as WiFi / power issue

### ISSUE THREAD

**Issue by idea--list** [sparkfunX/SparkFun_ESP32_Thing_Plus_C#6]
Recently i added WiFi functionality to my project and ever since then i am facing reliability issues.

The program logic (written in ESP-IDF) does collect sensor data and records it to a uSD card using the on-board slot. I also slice up the files into 1 minute long units. Once the file of the following minute is created and recorded to, i wait 10 seconds before transferring the previous file over wifi. The wifi transaction takes only 2-8 seconds each minute. The code runs without any issues while powered from USB cable, no matter how many days the session takes. I am also using the auto light sleep feature of ESP-IDF to keep the consumption as low as possible.

However as soon as i power the board from a single 18650 battery, the program stops running between the 50th and 60th minute and on the server side i regularly find the last file smaller then on the SD card, meaning the ESP32 stopped working right during transmission of the previous file (actually also my router could be the issue but i guess if that was true, i would face the same issue even when the board is powered via USB).

First i thought it might be due to the 28-30 AWGish wires on the battery holder. So i rewired it with 20 AWG pure copper wires (10 cm long). I measure 0.2 ohms resistance on my batt cables, which i consider low enough, the battery has 3000 mAh capacity and  should be able to deliver 1.5A current to the battery pins via 20 AWG. Tried several different cells, with same result. I am still facing the same issue, which is also hard to debug as if i connect the USB cable the issue is gone...

Checked the board design, you have 12 mil power traces which should be able to deliver upto 1A on outer layers supposing 1oz copper (though, the AP2112 will max out at 600mA).
I found many people facing similar issue on different forums, essentially they tell the ESP32 is becoming a power hungry beast when WiFi is being used. Some tell we should use beefy capacitors to avoid brownouts, some tell one should try LiFePO batteries to bypass voltage regulators that might not be capable to deliver enough power when the ESP has a power surge.

My question is: what should i try to debug the issue on this board? Where should i try to solder a cap of which capacitance? Is the 3V3 pin the right one to try a LiFePO4 battery to power the MCU directly? Do you have any other idea which may help solving the issue?

**Comment by PaulZC** [comment]
Hi Peter (@idea--list ),

We have a new version of the "Thing Plus C": https://www.sparkfun.com/products/20168 . It uses a newer 700mA 3.3V regulator, which may help. But I see that it does still use a 12mil power trace from the regulator to the ESP32 module.

My advice would be to try soldering a 100uF >=6V electrolytic capacitor between 3.3V and GND. That should help smooth out any power 'gulps' as the ESP32 transmits over WiFi.

If you have access to an oscilloscope, try probing on the 3.3V rail. Set the scope to trigger on a falling edge and set the trigger level to 3.2V. If you see the scope trigger, you know the voltage level is falling to a point where the regulator stops regulating. If it is a dual-channel scope, try using the second channel to monitor the battery voltage. Do you see the battery voltage dip at the same instant? What voltage does it dip to?

I hope this helps,
Paul

**Comment by idea--list** [comment]
Hello Paul (@PaulZC ),

Sorry for the delayed reply i have digged myself into my rabbit hole.
Thanks for the suggestions, i do not have an oscilloscope but i programmed another ESP32 to make voltage readings.
Could not notice anything special yet, but for sure it is not as good as an oscilloscope would be.

ESP-IDF has a ton of setting options in menukonfig and it seems after increasing different stack sizes, WiFi has become more stable... i am still crossing my fingers but i became cautious with WiFi on the ESP32 in general. Will need to run several long term tests in the following 1-2 weeks as it always takes a while until the problem occurs.

I seriously think you guys consider my whishes. I mean i was daydreaming about good SD slots and a DA16200 based feather and now you have both on offer ;-)

**Comment by idea--list** [comment]
@PaulZC 
Took long to sort it out, but i finally managed to solve it. 😄 
I was using edge type interrupts with my sensor and the on-board button. It turns out the ESP32 has a HW issue: the chip just stops triggering interrupts after a while (listed as error 3.14 in the [errata](https://www.espressif.com/sites/default/files/documentation/eco_and_workarounds_for_bugs_in_esp32_en.pdf)).

My sensor has an onboard FIFO, interrupt signal is issued when it is 50% full so that the MCU will readout the data. This readout also detects if the FIFO has been overflown or not. If so the sensor stops issuing further interrupt signals until its FIFO gets resetted. Exactly that is what happened because of the above error and the MCU got stuck in an infinite loop.

Until recently i did not pin my tasks to specific cores and was keep facing the same issue. After a ton trial&error it turns out that the workaround needs a core that can react instantly. However when the scheduler automatically manages which core will execute a call, then once every 1-3 hours the workaround can not be executed instantly and the same issue manifests as before applying the fix. It just takes even longer to sort it out...a

So long story short: WiFi and FTP logic can coexist just fine pinned to core 0, anything else needs to be on core 1 and we also need to be careful with task priorities to avoid priority inversions which would again reintroduce the same issue. Revised code seems to finish 6 hour long tests reliably without any capacitors. It was not a power supply issue as i supposed at the beginning.

**Comment by PaulZC** [comment]
Hi Peter (@idea--list ),

Wow. Great detective work!

Many thanks for the update. I'm going to pin this issue as other people will find it useful.

Very best wishes,
Paul
